airflow/models/taskinstance.py

```diff
@@ -1503,6 +1503,10 @@ def handle_failure_with_callback(
 
     def is_eligible_to_retry(self):
         """Is task instance is eligible for retry"""
+        if self.state == State.CLEARED_WHEN_RUNNING:
+            # If a task was cleared when running, it is always eligible for retry
```

Comment:
I am not sure you'd is the case. If it's already run its maximal number of tries out shouldn't run again should it?

Reply:
Hi @ashb , I think we should run it again. For example, if a task is configured with `max_retries=0`, and it's now running with current `try_number` 0. Then the user clears it (e.g. to pick up a code change), should we kill the job? Yes. Should it retry? I think the answer is Yes too. Intuitively, the Clear operation in Airflow is always associated with clearing the current state of the taskinstance and get it ready to run again once its dependencies are met. The only exception is if the task is currently running. If a running task is cleared, it currently fails, not because of the clearing, but because it gets a SIGTERM after its state is set to SHUTDOWN. So the failure and the `on_failure_callback` received when a running task is cleared is counter-intuitive to many users.

If the user insists he wants to see a failure, he can still use Mark Failed which does exactly that. So in order to make Clear consistently clear the state of the task, I think it makes more sense to quietly kill the task and let `is_eligible_to_retry` return True here. Let me know what you think.

Reply:
Okay yes, that is a compelling argument.

How about `RESTARTING` as the state name instead? It's slightly clearer what the behaviour will be (to me at least)

Though I guess this isn't quite accurate either, as it would go back via the scheduler loop, rather than a direct restart in the worker.)

Hmm

Reply:
Thanks @ashb. I agree `RESTARTING` is a more concise and easier to understand name. So I'm changing to that. 
Regarding "it would go back via the scheduler loop", I think that's actually a better behaviour. TaskInstances that are cleared should go back to the scheduler loop instead of being restarted by the worker immediately. One example scenario is a DAG like this:
```
A >> B
```
If A is in success state, and B is a long running task that's currently running. And some external conditions requires a user to re-run both A and B. He can do this by clearing A. This sets the state of A to NONE and state of B to RESTARTING. B is then sent a SIGTERM and goes into UP_FOR_RETRY. While that's happening, the scheduler starts to run A first. Once A finishes, B can then run. 

However, if we allow the worker to restart B immediately after it is cleared when running, it would have violated the dependency graph because A might not done yet when the worker restarts B.